app/src/main/java/pl/ppiwd/exerciseanalyst/activities/TrainingDataActivity.java

```diff
@@ -0,0 +1,254 @@
+        ArrayAdapter<String> adapter = new ArrayAdapter<>(TrainingDataActivity.this, android.R.layout.simple_spinner_item, activities);
+        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+        Spinner spinner = (Spinner) findViewById(R.id.activitySpinner);
+        spinner.setAdapter(adapter);
+        return spinner;
+    }
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_training);
+        initialize();
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        setResult(RESULT_CANCELED, null);
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        if (deviceConnectionChecker.isConnected()) {
+            bindToMetaMotionService();
+        }
+    }
+
+    @Override
+    protected void onStop() {
+        super.onStop();
+        finishSession();
+        deviceConnectionChecker.unregister();
+        serverConnection.cancelRequest();
+        Toast.makeText(this, "Unfinished trainings dropped", Toast.LENGTH_SHORT).show();
+    }
+
+    @Override
+    public void onBackPressed() {
+        if (startStopButton.getText().toString().equals(getString(R.string.finish_training))) {
+            Toast.makeText(getApplicationContext(), "Finish training before going back!", Toast.LENGTH_LONG).show();
+        } else {
+            finish();
+        }
+    }
+
+    private NumberPicker getPicker() {
+        NumberPicker picker = (NumberPicker) findViewById(R.id.repetitionsPicker);
+        picker.setMinValue(1);
+        picker.setMaxValue(25);
+        picker.setValue(5);
+        return picker;
+    }
+
+    private void handleSession() {
+        timer.reset();
+        timerHandler.post(timer);
+        startMetaMotionService();
+        startStopButton.setText(R.string.finish_training);
+        activitySpinner.setEnabled(false);
+        repetitionsPicker.setEnabled(false);
+        startStopButton.setOnClickListener(v -> {
+            timerHandler.removeCallbacks(timer);
+
+            finishSession();
+            sendDataToServer();
+        });
+    }
+
+    @NeedsPermission(Manifest.permission.ACCESS_FINE_LOCATION)
+    public void startMetaMotionService() {
+        if (deviceConnectionChecker.isConnected()) {
+            Toast.makeText(this, "Service is already running", Toast.LENGTH_SHORT).show();
+            return;
+        }
+
+        Toast.makeText(this, "Wait for connection with device...", Toast.LENGTH_LONG).show();
+
+        deviceService.putExtra(Constants.DEVICE_MAC_ADDRESS_KEY, getIntent().getStringExtra(Constants.DEVICE_MAC_ADDRESS_SHARED_PREFS_KEY));
+        deviceService.putExtra(Constants.ACTIVITY_NAME_KEY, activitySpinner.getSelectedItem().toString());
+        deviceService.putExtra(Constants.REPETITIONS_COUNT_KEY, repetitionsPicker.getValue());
+
+        startService(deviceService);
```

Comment:
please use  ContextCompat.startForegroundService(this, serviceIntent);  
I wouldn't change anything in activity <-> service communication at this point of the project. What we have is not perfect but it is tested and works reliably.